ThriftHandler/MapDHandler.cpp

```diff
@@ -1074,6 +1073,116 @@ void MapDHandler::load_table_binary_columnar(const TSessionId& session,
   loader->load(import_buffers, numRows);
 }
 
+using RecordBatchVector = std::vector<std::shared_ptr<arrow::RecordBatch>>;
+
+#define ARROW_THRIFT_THROW_NOT_OK(s)            \
+  do {                                          \
+    ::arrow::Status _s = (s);                   \
+    if (UNLIKELY(!_s.ok())) {                   \
+      TMapDException ex;                        \
+      ex.error_msg = _s.ToString();             \
+      LOG(ERROR) << s.ToString();               \
+      throw ex;                                 \
+    }                                           \
+  } while (0)
+
+static RecordBatchVector loadArrowStream(const std::string& stream) {
+  RecordBatchVector batches;
+  try {
+    // TODO(wesm): Make this simpler in general, see ARROW-1600
+    auto stream_buffer = std::make_shared<arrow::Buffer>(reinterpret_cast<const uint8_t*>(stream.c_str()),
+                                                         static_cast<int64_t>(stream.size()));
+
+    arrow::io::BufferReader buf_reader(stream_buffer);
+    std::shared_ptr<arrow::RecordBatchReader> batch_reader;
+    ARROW_THRIFT_THROW_NOT_OK(arrow::ipc::RecordBatchStreamReader::Open(&buf_reader, &batch_reader));
+
+    while (true) {
+      std::shared_ptr<arrow::RecordBatch> batch;
+      // Read batch (zero-copy) from the stream
+      ARROW_THRIFT_THROW_NOT_OK(batch_reader->ReadNext(&batch));
+      if (batch == nullptr) {
+        break;
+      }
+      batches.emplace_back(std::move(batch));
+    }
+  } catch (const std::exception& e) {
+    LOG(ERROR) << "Error parsing Arrow stream: " << e.what() << ". Import aborted";
+  }
+  return batches;
+}
+
+void MapDHandler::load_table_binary_arrow(const TSessionId& session,
+                                          const std::string& table_name,
+                                          const std::string& arrow_stream) {
+  check_read_only("load_table_binary_arrow");
+  const auto session_info = get_session(session);
+  auto& cat = session_info.get_catalog();
+  if (g_cluster && !leaf_aggregator_.leafCount()) {
+    // Sharded table rows need to be routed to the leaf by an aggregator.
+    check_table_not_sharded(cat, table_name);
+  }
+  const TableDescriptor* td = cat.getMetadataForTable(table_name);
+  if (td == nullptr) {
+    TMapDException ex;
+    ex.error_msg = "Table " + table_name + " does not exist.";
+    LOG(ERROR) << ex.error_msg;
+    throw ex;
+  }
+  check_table_load_privileges(session, table_name);
+  std::unique_ptr<Importer_NS::Loader> loader;
+  if (leaf_aggregator_.leafCount() > 0) {
+    loader.reset(new DistributedLoader(session_info, td, &leaf_aggregator_));
+  } else {
+    loader.reset(new Importer_NS::Loader(cat, td));
+  }
```

Comment:
We should probably extract the first ~20 lines of this method into a separate one since they do setup work and access checks which don't have anything to do with the format of the input data. I don't mind doing it myself after we land this, anyway.

Reply:
done